libraries/python/chat-driver/chat_driver/openai_chat_completion_driver.py

```diff
@@ -186,15 +224,16 @@ def update_instructions(self, instructions: str) -> None:
         the rest of the messages are preserved."""
 
         # If the first message is a system message, delete it:
-        if self.messages and self.messages[0]["role"] == "system":
-            self.messages = self.messages[1:]
+        messages = self.messages_provider.get()
```

Comment:
storing the system messages in the message history is weird. i think it's a mixing of concerns. and it requires that the message provider implement a handful of mutation functions (set, extend).

it seems that the "instructions" system message should be managed separately by the chat driver, and it should be prepended to the messages at chat-completion-call time

Reply:
Since different callers have different uses/abuses of system messages, passing all messages (system + assistant + user) and any number of each is a requirement for many of the use cases I would have for it within assistant development, so whatever changes are made, this is an important one to support - does that make it a concern of the messages provider?

Reply:
So, I left the system "instruction" behavior in here, but separated it like you suggested, @markwaddle . This is a convenience for people who want to do the "normal" thing of setting a single system message as an instruction (like in GPTs) which covers many use cases. However, since the message_provider is separate now, @bkrabach , you can do your more custom message managment, too... just don't provide `instructions` in the chat driver config.